### PR TITLE
Add delta rotation functions for orbital rotation

### PR DESCRIPTION
**Pull request by markdewing**
Add a function to compute the new rotation matrix (and corresponding parameters). 
This also needs a function to compute the indices for the full rotation matrix.

(At one time during the implementation of orbital rotation, it was important that the first part of the list of full rotation indices corresponded to the other list of rotation indices.  That might have been due to incorrect code.  I think with the correct code the ordering shouldn't matter.)

These will eventually support the global rotation method for orbital rotation optimization.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature
:

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
desktop

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by prckent** [comment]
Test this please

File: src/QMCWaveFunctions/RotatedSPOs.cpp [hunk 1]
```diff
@@ -40,6 +40,25 @@ void RotatedSPOs::createRotationIndices(int nel, int nmo, RotationIndices& rot_i
...
+
+  // Add core-core rotations - put them at the end of the list
+  for (int i = 0; i < nel; i++)
+    for (int j = i + 1; j < nel; j++)
+      rot_indices.emplace_back(i, j);
+
+  // Add active-active rotations - put them at the end of the list
+  for (int i = nel; i < nmo; i++)
+    for (int j = i + 1; j < nmo; j++)
+      rot_indices.emplace_back(i, j);
```
Comment (ye-luo) [inline src/QMCWaveFunctions/RotatedSPOs.cpp:59]: Is it possible to precompute the size of rot_indices and reserve it? I feel nmo * nel scale emplace_back can be quite costly due to resize.

File: src/QMCWaveFunctions/RotatedSPOs.cpp [hunk 2]
```diff
@@ -172,6 +191,41 @@ void RotatedSPOs::apply_rotation(const std::vector<RealType>& param, bool use_st
...
+  assert(old_param.size() == full_rot_inds.size());
+  assert(new_param.size() == full_rot_inds.size());
+
+  const size_t nmo = new_rot_mat.rows();
+  assert(new_rot_mat.rows() == new_rot_mat.cols());
+
+  ValueMatrix old_rot_mat(nmo, nmo);
+  old_rot_mat = ValueType(0);
+
+  constructAntiSymmetricMatrix(full_rot_inds, old_param, old_rot_mat);
```
Comment (ye-luo) [inline src/QMCWaveFunctions/RotatedSPOs.cpp:211]: Is the set 0 necessary? constructAntiSymmetricMatrix doesn't sound like need of setting zero before calling it.

File: src/QMCWaveFunctions/RotatedSPOs.cpp [hunk 3]
```diff
@@ -172,6 +191,41 @@ void RotatedSPOs::apply_rotation(const std::vector<RealType>& param, bool use_st
...
+  constructAntiSymmetricMatrix(act_rot_inds, delta_param, delta_rot_mat);
+  exponentiate_antisym_matrix(delta_rot_mat);
+
+  // Apply delta rotation to old rotation.
+  BLAS::gemm('N', 'N', nmo, nmo, nmo, 1.0, delta_rot_mat.data(), nmo, old_rot_mat.data(), nmo, 0.0, new_rot_mat.data(),
+             nmo);
+
+  ValueMatrix log_rot_mat(nmo, nmo);
+  log_rot_mat = new_rot_mat;
+  log_antisym_matrix(log_rot_mat);
```
Comment (ye-luo) [inline src/QMCWaveFunctions/RotatedSPOs.cpp:226]: Can this function do out-of-place transformation?

File: src/QMCWaveFunctions/RotatedSPOs.cpp (hunk 1 above)
Comment (markdewing) [reply src/QMCWaveFunctions/RotatedSPOs.cpp:59]: This function isn't going to run that frequently, and the resize algorithm should double the capacity every time it runs out, so this really shouldn't be a big issue.

But yes, the size can be precomputed, and it should be nmo*(nmo-1)/2.

File: src/QMCWaveFunctions/RotatedSPOs.cpp (hunk 2 above)
Comment (markdewing) [reply src/QMCWaveFunctions/RotatedSPOs.cpp:211]: constructAntiSymmetric matrix does need the input matrix to be zero.  Not all entries are referenced, and the unreferenced entries will not get set by the function.
It looks like ValueMatrix uses Vector as an underlying container and that will set the value to zero on initialization/resize.

File: src/QMCWaveFunctions/RotatedSPOs.cpp (hunk 3 above)
Comment (markdewing) [reply src/QMCWaveFunctions/RotatedSPOs.cpp:226]: It could be modified to do the transformation out-of-place.

**Comment by prckent** [comment]
Happy to merge this since getting to usable-for-science oo in the batch code is such a priority.

File: src/QMCWaveFunctions/RotatedSPOs.cpp (hunk 2 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/RotatedSPOs.cpp:211]: `unreferenced entries will not get set by the function.` that means `constructAntiSymmetric` assumes the previous value should be kept untouched. Is a desired safe behavior? I'm wondering if constructAntiSymmetric should set 0 internally. Whether ValueMatrix set 0 or not at initialization is not relevant when we define the behavior of constructAntiSymmetric I assume.

Comment (markdewing) [reply src/QMCWaveFunctions/RotatedSPOs.cpp:211]: There's a comment in constructAntiSymmetricMatrix that states the assumption that the rotation matrix is initialized to zero upon entry.   Based on your other review comments on this PR, I got the impression you were concerned about performance implications, and in this case it is redundantly setting the matrix to zero.    But constructAntiSymmetricMatrix could set the array to zero to ensure correctness.

Comment (ye-luo) [reply src/QMCWaveFunctions/RotatedSPOs.cpp:211]: In this case, correctness is favored.

**Comment by prckent** [comment]
Test this please

**Review by prckent — APPROVED** [review]
Concerns addressed.

Ye: Cancel the automerge if you have additional comments.

**Comment by ye-luo** [comment]
> Concerns addressed.
> 
> Ye: Cancel the automerge if you have additional comments.

@markdewing addressed all my concern.

**Pull request merged**